packages/type-utils/tests/isTypeReadonly.test.ts

```diff
@@ -0,0 +1,278 @@
+            );
+
+            const result = isTypeReadonly(checker, receiver, options);
+            expect(result).toBe(true);
+          });
+
+          it('handles a ReadonlyMap', () => {
+            const { type: receiver, checker } = getType(
+              `type Test = ReadonlyMap<string, string>;`,
+            );
+
+            const result = isTypeReadonly(checker, receiver, options);
+            expect(result).toBe(true);
+          });
+        });
+      });
+
+      describe('is not readonly', () => {
+        it('fails with record with mutable props', () => {
+          const { type: receiver, checker } = getType(
+            `type Test = { bar: string; };`,
+          );
+
+          const result = isTypeReadonly(checker, receiver);
+          expect(result).toBe(false);
+        });
+
+        it('fails with a mutable array', () => {
+          const { type: receiver, checker } = getType(`type Test = string[];`);
+
+          const result = isTypeReadonly(checker, receiver);
+          expect(result).toBe(false);
+        });
+
+        it('fails with a mutable Array', () => {
+          const { type: receiver, checker } = getType(
+            `type Test = Array<string>;`,
+          );
+
+          const result = isTypeReadonly(checker, receiver);
+          expect(result).toBe(false);
+        });
+
+        // Methods are mutable; hence failure.
+        it('fails with a ReadonlySet', () => {
+          const { type: receiver, checker } = getType(
+            `type Test = ReadonlySet<string>;`,
+          );
+
+          const result = isTypeReadonly(checker, receiver);
+          expect(result).toBe(false);
+        });
+
+        // Methods are mutable; hence failure.
+        it('fails with a ReadonlyMap', () => {
+          const { type: receiver, checker } = getType(
+            `type Test = ReadonlyMap<string, string>;`,
+          );
+
+          const result = isTypeReadonly(checker, receiver);
+          expect(result).toBe(false);
+        });
+      });
+    });
+
+    describe('Intersection', () => {
+      describe('is readonly', () => {
+        describe('default options', () => {
+          it('handles an intersection of 2 records', () => {
+            const { type, checker } = getType(
+              `type Test = Readonly<{ foo: string; bar: number; }> & Readonly<{ bar: number; }>;`,
+            );
+
+            const result = isTypeReadonly(checker, type);
+            expect(result).toBe(true);
+          });
+
+          it('handles a readonly intersection of readonly array and readonly record', () => {
+            const { type, checker } = getType(
+              `type Test = Readonly<ReadonlyArray<string> & { readonly foo: string; }>;`,
+            );
+
+            const result = isTypeReadonly(checker, type);
+            expect(result).toBe(true);
+          });
+
+          it('handles a intersection of readonly readonly array and readonly record', () => {
```

Comment:
To get this test to pass, we may need to set up custom logic explicitly for intersection rather than relying on the general `isTypeReadonlyObject` check.

Doing so would also allow for treating arrays as a special case of object in intersection just like they are treated elsewhere (readonly arrays don't have to have readonly methods to be consider readonly like all other objects do).

Note: If extending an interface isn't internally treated the same as an intersection; then logic for that will have to be added too.

I'll look into this more tomorrow.